README.md
# XarigamiND TinyMCE


XarTinyMCE is based on TinyMCE (a very old version), a platform independent, web based Javascript HTML WYSIWYG 
editor released as Open Source under LGPL by Moxiecode Systems AB

xarTinyMCE has a number of nice features such as browser support for all modern browsers 
(currently Mozilla, MSIE, FireFox, Opera and Safari) and support for XHTML 1.0 strict markup. 
There are powerful features for not only specifying which xhtml tags are allowed, but also which
attributes and attribute values. You can enhance xarTinyMCE further with plugins and different configuration settings.

Original Xarigami project page: http://xarigami.org/project/xartinymce


xardocs/changelog.txt
* Version 3.4.0 5 April 2013
- Update to upstream git revision 614cff5 in 3.5.8 branch

* Version 3.3.0 5 May 2012
- Update to upstream git revision b81073c in 3.5.1 branch

* Version 3.3.0.b1 15 April 2012
- Update to upstream git revision 8457778 in 3.5.0 (beta4) branch
- Remove the xar-form-container* divs around the wysiwyg output for better/easier templating
- Update plugins
- This is a major upstream upgrade with added support for HTML5, rewritten plugin for non-editable content,
styles preview in dropdown.

* Version 3.2.5 Nov 11th 2011
- Update to upstream Tinymce Version 3.4.7

* Version 3.2.4 September 11th 2011
- Update to upstream Tinymce Version 3.4.5 (rev 3144ac76f8623b4bbfba9b1b58d979fa2881a5e8)

* Version 3.2.3 September 1st 2011
Fix plugin configuration for iBrowser:
 - Fix the dynamic dir default to /var/uploads/images/
 - Use private var directory (if available) for derivative images
 - Use images module derivative store if available instead of var/templates/
 - Fix typo in update config

* Version 3.2.2 August 28 2011
 - Update to upstream Tinymce Version 3.4.4
 - Bug fixes as per upstream project. Many related to style tag, list tag, HTML5 enhancements.

* Version 3.2.1 July 10 2011
 - Update to upstream Tinymce Version 3.4.3.2
 - Updated spell checker

* Version 3.2.0 April 11 2011
 - Update to upstream Tinymce Version 3.4.2
 - xtm-000811   Bug    Bad instance in security check 12 Apr 11 Normal 0% Major
 - xtm-000805   Improvement    Add new privillege mask SubmitTinyMCE
 - xtm-000803   New feature     Added configurations mechamism for plugins (eg ibrowser)

* Version 3.1.4 March 11 2011
 - Update to upstream Tinymce Version 3.4
 - Add aggregate="false" to tinymce so it is not aggregated with other js when aggregate is on

* Version 3.1.3 February 20 2011
- Update to reflect changes in WYSIWYG property for new validations in Dynamic data

* Version 3.1.2 25 Sept 2010
- Update to upstream Tinymce Version 3.3.9.1
- Fixes for IE6 beta
- Bug fixes from upstream project
- Updates to dphighlight syntax highlight plugin (must have dphighlight module installed)

* Version 3.1.1  May 2010
- Update to upstream TinyMCE version 3.3.6
- Update documentation
- Note updated key plugins iBrowser and Imanager

* Version 3.1.0 12 May 2010
- Update to upstream TinyMCE version 3.3.5.1
  This fixes many issues related to the table plugin, removing formats in Gecko browsers,
  IE DOM, nested format problems and more
- Cosmetic updates including updated icons,
- Breadcrumb links and menus updated to current styles and breadcrumb enhancements
- Liststyle plugin has been deprecated : use the advlist plugin instead.

* Version 3.0.4 10 Mar 2010
xtm-000527 Can't Check "Show advanced options?
- Other small bug fixes

* Version 3.0.3  23 Feb 2010
xtm-000502 Modify config template checkbox list tag
xtm-000513 Update code to upstream version 3.3.b
xtm-000517 Manual additon of template snippet for specific config
xtm-000518 Update with advlist plugin
- Bug fixes from the upstream tinymce project esp related to IE
- Much new functionality from the upstream tinymce project including
Paste plugin improved functionality, new text formatting engine among others.

* Version 3.0.2  09 Jan 2010
- Fix the autoloading on all textareas (unless mceNoEditor is the class)
- Update code to upstream version 3.2.7
- Add server side validation on creation of new wysiwyg instance
- Tidy up the UI and menus


* Version 3.0.1   18 Dec 2009
- Add autoloading option
- Fix undeclared vars
- Rework on organisation of property code
- Ensure we can override selector in template
- Rework some aspects of the UI for clearer presentation

* Version 3.0.0   20 Sept 2009
- First release in this series
- Incorporates changes from upstream project to version 3.2.6
- total rewrite to better support multiple configs
- rewritten user interface
- not backward compatible - requires an uninstall and reinistall


* Version 1.6.0 6 Aug 2009
- The last version for the 1x series
- Added upstream tinymce code to release 3.2.5
- A number of major bug fixes in the upstream project covering
    Fixes for Webkit bugs related to tables, and DOM ranges
    New autoresize options
    Better support for jquery
    Specific IE fixes for memory leaks and other issues
    Others as per the upstream project change log.
- Local bug fixes include:
    xtm-000018  Bug     Popup windows are not resizeable
    xtm-000391  Bug     Editor UI CSS File results in load issues?
    xtm-000426  Bug     Deprecated vars causing problems
    xtm-000427  Bug     Default class overrides GUI assignable
    xtm-000425  Improvement Update documentation

* Version 1.5.7
 - Added upstream tinymce code to release 3.2.4.1 and upped version to 1.5.7 to reflect

* Version 1.5.6  31 May 2009
- Applied changes from upstream 3.2.3.1 release.
See the tinymce changelog file in the modules/tinymce/xarincludes folder for full details

* Version 1.5.6  31 May 2009
- Applied changes from upstream 3.2.4.1 release.
See the tinymce changelog file in the modules/tinymce/xarincludes folder for full details.

* Version 1.5.5  5 May 2009
- Applies changes from upstream 3.2.3.1 release with fixes including:
- Fixed - paragraphs containing form elements such as input or textarea would be removed.
- Fixed - some IE versions would produce a wrapper function for events attributes.
- Fixed - table cell contents could be removed if you pressed return/enter at the end of the cell contents.
- Fixed - the paste plugin would remove a extra character if the selection range was collapsed.
- Fixed - creating tables with % width wouldn't be handled correctly on WebKit browsers.


* Version 1.5.5  30 April 2009
- Upgrade to latest upstream Revision 1075 including 3.2.3 release code
  Fixes a number of key bugs for IE
  See the upstream tinymce changelog for full details.

* Version 1.5.4  4 April 2009
- Upgrade to latest upstream Revision 1075 including 3.2.2.3 release code

xarinit.php
<?php
/**
 * xarTinyMCE initialization
 *
 * @package modules
 * @copyright (C) 2004-2013 2skies.com
 * @license GPL {@link http://www.gnu.org/licenses/gpl.html}
 * @link http://xarigami.com/project/xartinymce
 *
 * @subpackage xartinymce module
 * @author Jo Dalle Nogare <icedlava@2skies.com>
 */
/**
 * Initialise the tinymce module
 *
 * @access public
 * @param none
 * @returns bool true
 */
function tinymce_init()
{
    xarDBLoadTableMaintenanceAPI();
    $dbconn = xarDBGetConn();
    $xartables = xarDBGetTables();

    $tinymcetable = $xartables['tinymce'];
    //create the configuration settings table
    $fields = array('xar_id'        => array('type' => 'integer', 'null' => false,  'increment' => true, 'primary_key' => true),
                    'xar_name'      => array('type' => 'varchar', 'size' => 32,     'null' => false,     'default'=>''),
                    'xar_desc'      => array('type' => 'varchar', 'size' => 255,    'null' => false,     'default'=>''),
                    'xar_jsstring'  => array('type' => 'text',    'size'=>'medium'),
                    'xar_gzstring'  => array('type' => 'text',    'size'=>'medium'),
                    'xar_conftype'  => array('type' => 'integer', 'size'=>'small',  'default'=>'0'),
                    'xar_options'   => array('type' => 'text',    'size'=>'medium'),
                    'xar_active'    => array('type' => 'integer', 'size'=>'small', 'default'=>'0')
    );
    $query = xarDBCreateTable($tinymcetable,$fields);
    if (empty($query)) return; // throw back

    // Pass the Table Create DDL to adodb to create the table and send exception if unsuccessful
    $result = $dbconn->Execute($query);
    if (!$result) return;

    $baseurl = xarServerGetBaseURL();
    $baselength = strlen($baseurl);
    $default ='a:22:{s:5:"iname";s:7:"default";s:4:"desc";s:21:"Default configuration";s:6:"active";b:1;s:9:"useswitch";b:1;s:8:"autoload";b:1;s:6:"usegzp";b:1;s:8:"browsers";s:23:"msie,gecko,opera,safari";s:17:"editor_deselector";s:11:"mceNoEditor";s:4:"mode";s:9:"textareas";s:5:"theme";s:8:"advanced";s:8:"language";s:2:"en";s:7:"plugins";a:12:{i:0;s:8:"advimage";i:1;s:7:"advlink";i:2;s:7:"advlist";i:3;s:8:"emotions";i:4;s:10:"fullscreen";i:5;s:10:"loremipsum";i:6;s:9:"pagebreak";i:7;s:5:"paste";i:8;s:5:"print";i:9;s:13:"searchreplace";i:10;s:12:"spellchecker";i:11;s:5:"table";}s:17:"document_base_url";s:'.$baselength.':"'.$baseurl.'";s:16:"invalid_elements";s:0:"";s:27:"theme_advanced_blockformats";s:58:"p,address,pre,h1,h2,h3,h4,h5,h6,blockquote,dt,dd,code,samp";s:31:"theme_advanced_toolbar_location";s:3:"top";s:33:"theme_advanced_statusbar_location";s:6:"bottom";s:27:"theme_advanced_buttons1_add";s:47:"search,replace,pastetext,pasteword,spellchecker";s:27:"theme_advanced_buttons2_add";s:35:"print,fullscreen,emotions,pagebreak";s:27:"theme_advanced_buttons3_add";s:24:"tablecontrols,loremipsum";s:23:"theme_advanced_resizing";b:1;s:6:"custom";s:0:"";}';

    $jsstring = '
browsers: "msie,gecko,opera,safari",
editor_deselector: "mceNoEditor",
mode: "textareas",
theme: "advanced",
plugins: "advimage,advlink,advlist,emotions,fullscreen,loremipsum,pagebreak,paste,print,searchreplace,spellchecker,table",
invalid_elements: "",
theme_advanced_blockformats: "p,address,pre,h1,h2,h3,h4,h5,h6,blockquote,dt,dd,code,samp",
theme_advanced_toolbar_location: "top",
theme_advanced_statusbar_location: "bottom",
theme_advanced_buttons1_add: "search,replace,pastetext,pasteword,spellchecker",
theme_advanced_buttons2_add: "print,fullscreen,emotions,pagebreak",
theme_advanced_buttons3_add: "tablecontrols,loremipsum",
theme_advanced_resizing: true,
language: "en"';
    $jsstring = addslashes($jsstring);
    $gzstring ='
theme : "advanced",
plugins : "advimage,advlink,advlist,emotions,fullscreen,loremipsum,pagebreak,paste,print,searchreplace,spellchecker,table",
disk_cache : true,
language :  "en"';

    $gzstring = addslashes($gzstring);
    //insert a default instance config
    $nextId = $dbconn->GenId($tinymcetable);
    $query = "INSERT INTO $tinymcetable
                         (xar_id, xar_name,xar_desc, xar_jsstring, xar_gzstring, xar_conftype, xar_options, xar_active)
                  VALUES (?, 'default','Default tinymce configuration','$jsstring','$gzstring',0,'$default',1)";
    $result = $dbconn->Execute($query,array($nextId));
    if (!isset($result)) return;

    /* Set default module vars */
    xarModSetVar('tinymce', 'activetinymce',false);
    xarModSetVar('tinymce', 'default','default');
    xarModSetVar('tinymce', 'tinyloadmode','auto');
    xarModSetVar('tinymce', 'itemsperpage', 30);
    xarModSetVar('tinymce', 'useadvanced', false);
    xarModSetVar('tinymce','defaultconfig',$default);//original default
    //setup the main array of config values
    include 'modules/tinymce/xarincludes/configarray.php';

    $configs = serialize($configs);
    xarModSetVar('tinymce','configs',$configs);

    //setup the main array of button values
    include 'modules/tinymce/xarincludes/buttonarray.php';

    $buttons = serialize($buttonarray);
    xarModSetVar('tinymce','buttons',$buttons);

    /* let's set a default button switch  */
    $buttonon = xarML('Turn On');
    $buttonoff = xarML('Turn Off');
    $buttonswitch  = '
        function mce_button_toggle(form_element_id, button_o)
        {
            if (tinyMCE.activeEditor == null) {
                load_tinymce();
                button_o.value = "'.$buttonoff.'";
            } else {
                if (tinyMCE.get(form_element_id).isHidden()) {
                    tinyMCE.get(form_element_id).show();
                    button_o.value = "'.$buttonoff.'";
                } else {
                    tinyMCE.get(form_element_id).hide();
                    button_o.value = "'.$buttonon.'";
                }
            }
          return false;
        }';
    xarModSetvar('tinymce','buttonstring',$buttonswitch);

    //Define instances for masks
    $query1 = "SELECT DISTINCT xar_id, xar_name FROM  $tinymcetable";
            $instances = array(
                        array('header' => 'Instance ID:',
                                'query' => $query1,
                                'limit' => 20
                            )
                    );
            xarDefineInstance('tinymce', 'Instance', $instances);

    /* Set masks */
    xarRegisterMask('ViewTinyMCE','All','tinymce','instance','All','ACCESS_OVERVIEW');
    xarRegisterMask('ReadTinyMCE','All','tinymce','instance','All','ACCESS_READ');
    xarRegisterMask('EditTinyMCE','All','tinymce','instance','All','ACCESS_EDIT');
    xarRegisterMask('AddTinyMCE','All','tinymce','instance','All','ACCESS_ADD');
    xarRegisterMask('DeleteTinyMCE ','All','tinymce','instance','All','ACCESS_DELETE');
    xarRegisterMask('AdminTinyMCE','All','tinymce','instance','All','ACCESS_ADMIN');

    return tinymce_upgrade('3.0.0');
}

/**
 * Separate activation routines if necessary
 *
 * @access public
 * @param none $
 * @returns bool
 */
function tinymce_activate()
{
    /* Activate successful */
    return true;
}

/**
 * Upgrade the tinymce module from an old version
 *
 * @access public
 * @param oldVersion $
 * @returns bool
 * @raise DATABASE_ERROR
 */
function tinymce_upgrade($oldversion)
{

    switch ($oldversion) {
        // Upgrading from 3.0.0
        case '3.0.0':

            /*
            // BETA version upgrade.
            // Add the active field to the database table
            $dbconn = xarDBGetConn();
            $xartables = xarDBGetTables();
            $datadict = xarDBNewDataDict($dbconn, 'ALTERTABLE');
            $tinymcetable = $xartables['tinymce'];
            $active_field = 'xar_active SMALLINT DEFAULT 0';
            $result = $datadict->addColumn($tinymcetable, $active_field);
            if (!$result) return FALSE;

            // Assume all current instances are active...
            $iconfigs = xarModAPIFunc('tinymce', 'user', 'getall');
            foreach ($iconfigs as $name => $options) {
                $options['active'] = 1;
                if (!xarModAPIFunc('tinymce', 'admin', 'update', $options)) return false;
            }
            */

            // Update the config array to include autoloading
            include 'modules/tinymce/xarincludes/configarray.php';
            $configs = serialize($configs);
            xarModSetVar('tinymce','configs',$configs);

            // The toggle button needs to run the init method if TinyMCE hasn't been initialized on page load
            $buttonon = xarML('Turn On');
            $buttonoff = xarML('Turn Off');
            $buttonswitch  = '
                function mce_button_toggle(form_element_id, button_o)
                {
                    if (tinyMCE.activeEditor == null) {
                        load_tinymce();
                        button_o.value = "'.$buttonoff.'";
                    } else {
                        if (tinyMCE.get(form_element_id).isHidden()) {
                            tinyMCE.get(form_element_id).show();
                            button_o.value = "'.$buttonoff.'";
                        } else {
                            tinyMCE.get(form_element_id).hide();
                            button_o.value = "'.$buttonon.'";
                        }
                    }
                  return false;
                }';
            xarModSetvar('tinymce','buttonstring',$buttonswitch);

        case '3.0.1':
        case '3.0.2':
        case '3.0.3': // version update to reflect updated code, no db changes
            xarModAPIFunc('tinymce','admin','resetdefault',array('resettype'=>'configs'));
        case '3.0.4': // version update to reflect updated code, no db changes
            xarModAPIFunc('tinymce','admin','resetdefault',array('resettype'=>'configs'));
        case '3.1.0': // version update to reflect updated code, and stability level for this V 3 series
        case '3.1.1':
            //update to reflect upstream project fixes added to 3.3.9.1
            //plugin fixes (dphighlight)
        case '3.1.2':
            //update to reflect changes for new property validations
        case '3.1.3':
            //upgrade to Unstream tinymce 3.4
        case '3.1.4':
           xarRegisterMask('SubmitTinyMCE','All','tinymce','instance','All','ACCESS_COMMENT');
        case '3.2.0': //current version
            //upgraded to 3.4.2 plus xarigami 1.4.0 only
        case '3.3.0':
            //upgraded to 3.5.1
         case '3.4.0':
            //upgraded to 3.5.8 upstream version
        default:
       break;
       //always reset the configs in case there are new or changed ones
        xarModAPIFunc('tinymce','admin','resetdefault',array('resettype'=>'configs'));
    }
    return true;
}

/**
 * Delete the tinymce module
 *
 * @access public
 * @param none $
 * @returns bool true
 */
function tinymce_delete()
{
    xarDBLoadTableMaintenanceAPI();
    $dbconn = xarDBGetConn();
    $xartables = xarDBGetTables();
    $tinymcetable = $xartables['tinymce'];

    // Generate the SQL to drop the table using the API
     $query = xarDBDropTable($tinymcetable);
    try {
          $result = $dbconn->Execute($query);
    } catch (Exception $e) {
        //
    }
    //if (empty($query)) return; // throw back

    //if (!$result) return
    xarModDelAllVars('tinymce');

    // Remove Masks and Instances
    xarRemoveMasks('tinymce');
    xarRemoveInstances('tinymce');
    return true;
}

?>